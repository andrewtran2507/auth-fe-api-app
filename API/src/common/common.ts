import * as argon2 from 'argon2';

// export const hashData = (data: string) => {
//     return argon2.hash(data);
// }

export function hashData(data: string) {
  return argon2.hash(data);
}
