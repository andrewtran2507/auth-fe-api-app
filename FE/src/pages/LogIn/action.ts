import axios from '../../utils/axios';

export const getUser = async (email: string | null, password: string | null) => {
  const response = await axios.post(
    'auth/login',
    JSON.stringify({
      email,
      password,
    }),
  );
  return response.data;
};
