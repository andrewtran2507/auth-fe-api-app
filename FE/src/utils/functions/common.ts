export const handleFormLogInValidate = (
  email: string | null,
  password: string | null,
  username?: string | null,
) => {
  const passwordRegex = /^.{7,}$/;
  const emailRegex = /^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$/;
  let emailError = '';
  let passwordError = '';
  let usernameError = '';
  if (email === '') {
    emailError = 'Email is required';
  } else if (email && !emailRegex.test(email)) {
    emailError = 'Email is invalid';
  }
  if (password === '') {
    passwordError = 'Password is required';
  } else if (password && !passwordRegex.test(password)) {
    passwordError = 'Password should contain at least 7 characters';
  }
  if (username === '') {
    usernameError = 'Username is required';
  }
  return {
    emailError,
    passwordError,
    usernameError,
  };
};
