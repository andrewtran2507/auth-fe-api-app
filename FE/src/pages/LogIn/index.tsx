import React, { FC, useState, useEffect, useMemo } from 'react';
import { useNavigate } from 'react-router-dom';
import { AxiosError } from 'axios';
import { useQuery } from 'react-query';

// MUI
import Avatar from '@mui/material/Avatar';
import CssBaseline from '@mui/material/CssBaseline';
import TextField from '@mui/material/TextField';
import Box from '@mui/material/Box';
import LockOutlinedIcon from '@mui/icons-material/LockOutlined';
import Typography from '@mui/material/Typography';
import Container from '@mui/material/Container';
import { Alert } from '@mui/material';
import ReportIcon from '@mui/icons-material/Report';
import CheckCircleIcon from '@mui/icons-material/CheckCircle';
import { LoadingButton } from '@mui/lab';

import { handleFormLogInValidate } from '../../utils/functions/common';
import { EIconStatus, TLoginInput, TResponseError } from '../../types';
import { getUser } from './action';
import './index.scss';

const LogIn: FC = () => {
  const navigate = useNavigate();

  const [formValue, setFormValue] = useState<TLoginInput>({ email: '', password: '' });
  const [formError, setFormError] = useState<TLoginInput>({ email: '', password: '' });
  const [isSubmit, setIsSubmit] = useState<boolean>(false);
  const [message, setMessage] = useState<string>('');
  const [iconStatus, setIconStatus] = useState<
    'success' | 'info' | 'warning' | 'error' | undefined
  >(EIconStatus.success);

  const getUserAction = async () => await getUser(formValue.email, formValue.password);
  const { data, error, refetch, isLoading } = useQuery('getUser', getUserAction, {
    enabled: false,
  });

  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    setIsSubmit(true);
    const formData = new FormData(event.currentTarget);
    const email = formData.get('email') as string | null;
    const password = formData.get('password') as string | null;
    setFormValue({ email, password });

    const check = handleFormValidate(email, password);
    if (!check) {
      return;
    }
    refetch();
  };
  const handleFormValidate = (email: string | null, password: string | null) => {
    const validateResult = handleFormLogInValidate(email, password);
    setFormError({ email: validateResult.emailError, password: validateResult.passwordError });
    if (Object.values(validateResult).some((value) => value !== '')) {
      return false;
    }
    return true;
  };

  useEffect(() => {
    if (data) {
      setMessage('Login successfully');
      setIconStatus(EIconStatus.success);
      localStorage.setItem('auth', JSON.stringify(data));
      navigate(0);
    }
    return;
  }, [data]);

  useEffect(() => {
    if (error) {
      setIconStatus(EIconStatus.error);
      const errorLogin = error as AxiosError;
      if (errorLogin.response) {
        const res = errorLogin.response.data as TResponseError;
        setMessage(res.message);
      } else if (errorLogin.request) {
        setMessage('Network Error');
      } else {
        setMessage('Something went wrong');
      }
    }
  }, [error]);

  useEffect(() => {
    setMessage('');
    isSubmit && handleFormValidate(formValue.email, formValue.password);
  }, [formValue, isSubmit]);
  useEffect(() => {
    setMessage('');
  }, [formValue]);

  const handleNavigateSignUp = () => {
    navigate('/sign-up');
  };

  const IconStatusComponent =
    iconStatus === EIconStatus.success ? <CheckCircleIcon /> : <ReportIcon />;

  const messageVisibility = useMemo(() => {
    return message === '' ? 'hidden' : 'visible';
  }, [message]);

  return (
    <>
      <div className="login">
        <Container component="main" maxWidth="xs" className="container">
          <CssBaseline>
            <Box
              sx={{
                display: 'flex',
                flexDirection: 'column',
                alignItems: 'center',
                padding: '1.5em 1.5em',
                borderRadius: 1,
                border: '1px solid',
                borderColor: '#b0bec5',
              }}>
              <Avatar sx={{ m: 1, bgcolor: 'secondary.main' }}>
                <LockOutlinedIcon />
              </Avatar>
              <Typography component="h1" variant="h5">
                Login
              </Typography>
              <Box
                component="form"
                onSubmit={handleSubmit}
                noValidate
                sx={{ mt: 1 }}
                className="form">
                <TextField
                  inputProps={{ 'data-testid': 'email-input' }}
                  margin="normal"
                  required
                  fullWidth
                  id="email"
                  label="Email Address"
                  name="email"
                  autoComplete="email"
                  autoFocus
                  helperText={formError.email}
                  error={isSubmit && formError.email !== ''}
                  onChange={(value) =>
                    setFormValue((state) => {
                      return { ...state, email: value.target.value };
                    })
                  }
                />
                <TextField
                  inputProps={{ 'data-testid': 'password-input' }}
                  margin="normal"
                  required
                  fullWidth
                  name="password"
                  label="Password"
                  type="password"
                  id="password"
                  autoComplete="current-password"
                  helperText={formError.password}
                  error={isSubmit && formError.password !== ''}
                  onChange={(value) =>
                    setFormValue((state) => {
                      return { ...state, password: value.target.value };
                    })
                  }
                />
                <div style={{ visibility: messageVisibility }}>
                  <Alert color={iconStatus} icon={IconStatusComponent}>
                    {message}
                  </Alert>
                </div>
                <LoadingButton
                  loading={isLoading}
                  type="submit"
                  fullWidth
                  variant="contained"
                  sx={{ mt: 3, mb: 2 }}>
                  Login
                </LoadingButton>
                <div className="bottom-text">
                  <span>Doesn&apos;t have an account?</span>
                  <span onClick={handleNavigateSignUp} className="navigate-text">
                    Signup
                  </span>
                </div>
              </Box>
            </Box>
          </CssBaseline>
        </Container>
      </div>
    </>
  );
};

export default LogIn;
