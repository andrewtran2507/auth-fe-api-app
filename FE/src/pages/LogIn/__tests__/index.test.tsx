import {
  render,
  screen,
  act,
  RenderResult,
  Queries,
  waitFor,
  cleanup,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import LogIn from '..';
import { WrapperQueryClient as wrapper } from '../../../utils/test/common';

const mockUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockUsedNavigate,
}));

describe('Login form', () => {
  let component: RenderResult<Queries, HTMLElement, HTMLElement> | undefined = undefined;
  beforeEach(() => {
    jest.clearAllMocks();
  });
  afterEach(() => {
    cleanup();
  });

  it('Test to match snapshot of component', async () => {
    await act(() => {
      component = render(<LogIn />, { wrapper });
    });
    const { asFragment } = component || { asFragment: () => DocumentFragment };
    expect(asFragment()).toMatchSnapshot();
  });

  test('TextField email address receives value that user is typing into', async () => {
    await act(async () => render(<LogIn />, { wrapper }));
    const input = await screen.findByTestId('email-input');
    await waitFor(() => {
      userEvent.type(input, 'Jay');
      expect(input).toHaveValue('Jay');
    });
  });

  test('TextField password receives value that user is typing into', async () => {
    await act(async () => render(<LogIn />, { wrapper }));
    const input = await screen.findByTestId('password-input');
    await waitFor(() => {
      userEvent.type(input, 'secret007');
      expect(input).toHaveValue('secret007');
    });
  });

  test('Button Login has name as Login', async () => {
    await act(async () => render(<LogIn />, { wrapper }));
    const buttonSubmit = await screen.findByRole('button', { name: 'Login' });

    await waitFor(() => {
      expect(buttonSubmit).toBeInTheDocument();
    });
  });
});
