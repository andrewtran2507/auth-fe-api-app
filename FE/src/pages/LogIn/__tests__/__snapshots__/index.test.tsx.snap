// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login form Test to match snapshot of component 1`] = `
<DocumentFragment>
  <div
    class="login"
  >
    <main
      class="MuiContainer-root MuiContainer-maxWidthXs container css-nzinn5-MuiContainer-root"
    >
      <div
        class="MuiBox-root css-ojx66d"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1999axt-MuiAvatar-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="LockOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
            />
          </svg>
        </div>
        <h1
          class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          Login
        </h1>
        <form
          class="form MuiBox-root css-164r41r"
          novalidate=""
        >
          <div
            class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-focused Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for="email"
              id="email-label"
            >
              Email Address
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                autocomplete="email"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                data-testid="email-input"
                id="email"
                name="email"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    Email Address *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="password"
              id="password-label"
            >
              Password
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                autocomplete="current-password"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                data-testid="password-input"
                id="password"
                name="password"
                required=""
                type="password"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Password *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            style="visibility: hidden;"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardSuccess MuiAlert-standard css-wnkddu-MuiPaper-root-MuiAlert-root"
              role="alert"
            >
              <div
                class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="CheckCircleIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                  />
                </svg>
              </div>
              <div
                class="MuiAlert-message css-1pxa9xg-MuiAlert-message"
              />
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1qudo6r-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
            id=":r2:"
            tabindex="0"
            type="submit"
          >
            Login
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <div
            class="bottom-text"
          >
            <span>
              Doesn't have an account?
            </span>
            <span
              class="navigate-text"
            >
              Signup
            </span>
          </div>
        </form>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
